File: apps/desktop/src/services/sessionSearch.ts
/**
 * 会话搜索：按标题搜，覆盖项目下任务 + 收集箱对话。
 *
 * 搜索策略：标题子串命中优先，同时用字符 bigram TF-IDF + 余弦相似度补充相近结果；
 * 按 route 合并去重后排序。
 *
 * 当前只搜标题——消息体存在内存 HashMap 重启就丢，等持久化落盘后再扩到 content。
 */

import { listProjects } from "./projectsStore";
import {
  listOrphanConversations,
  listProjectConversations,
} from "./tasksStore";
import {
  areSidebarConversationsLoaded,
  ensureSidebarConversationsLoaded,
  getSidebarConversationRevision,
  listSidebarConversations,
} from "./sidebarConversations";

export type SearchKind = "project-task" | "orphan";

export interface SearchResult {
  kind: SearchKind;
  /** 项目任务才有；orphan 时为 undefined。 */
  projectId?: string;
  /** 项目展示名；orphan 时为 undefined（UI 显示「收集箱」标签）。 */
  projectName?: string;
  taskId: string;
  title: string;
  /** 直接可以 router.push 的路径。 */
  route: string;
  /** 越大越相关。不同模式的量纲不同，UI 只用来排序。 */
  score: number;
  /** title 上需要高亮的 [start, end) 区间。纯向量命中时为空。 */
  highlights: Array<[number, number]>;
}

interface Doc {
  kind: SearchKind;
  projectId?: string;
  projectName?: string;
  taskId: string;
  title: string;
  route: string;
}

interface IndexedDoc extends Doc {
  titleTokens: string[];
  vector: Map<string, number>;
}

interface IndexedCorpus {
  key: string;
  docs: IndexedDoc[];
  idf: Map<string, number>;
}

let indexedSidebarCorpus: IndexedCorpus | null = null;

function buildStoreCorpus(): Doc[] {
  const docs: Doc[] = [];
  for (const p of listProjects()) {
    for (const t of listProjectConversations(p.id)) {
      docs.push({
        kind: "project-task",
        projectId: p.id,
        projectName: p.name,
        taskId: t.id,
        title: t.title,
        route: `/projects/${p.id}/tasks/${t.id}`,
      });
    }
  }
  for (const o of listOrphanConversations()) {
    docs.push({
      kind: "orphan",
      taskId: o.id,
      title: o.title,
      route: `/chats/${o.id}`,
    });
  }
  return docs;
}

function buildSidebarCorpus(): Doc[] {
  return listSidebarConversations().map((conversation) => ({
    kind: conversation.projectId ? "project-task" : "orphan",
    projectId: conversation.projectId ?? undefined,
    projectName: conversation.projectName ?? undefined,
    taskId: conversation.taskId,
    title: conversation.title,
    route: conversation.route,
  }));
}

function buildSidebarCorpusKey(): string | null {
  if (!areSidebarConversationsLoaded()) return null;
  return `sidebar:${getSidebarConversationRevision()}`;
}

// ---------------- text mode ----------------

/** 同一查询在标题里所有命中点的 [start, end) 区间。 */
function findRanges(title: string, query: string): Array<[number, number]> {
  const t = title.toLowerCase();
  const q = query.toLowerCase().trim();
  if (!q) return [];
  const ranges: Array<[number, number]> = [];
  let idx = 0;
  while ((idx = t.indexOf(q, idx)) !== -1) {
    ranges.push([idx, idx + q.length]);
    idx += q.length;
  }
  if (ranges.length > 0) return ranges;
  // 整串搜不到 → 按 whitespace 拆 token 再分别找。
  for (const tok of q.split(/\s+/).filter(Boolean)) {
    let i = 0;
    while ((i = t.indexOf(tok, i)) !== -1) {
      ranges.push([i, i + tok.length]);
      i += tok.length;
    }
  }
  return ranges;
}

function toSearchResult(doc: Doc, score: number, highlights: Array<[number, number]>): SearchResult {
  return {
    kind: doc.kind,
    projectId: doc.projectId,
    projectName: doc.projectName,
    taskId: doc.taskId,
    title: doc.title,
    route: doc.route,
    score,
    highlights,
  };
}

function searchText(query: string, corpus: readonly Doc[]): SearchResult[] {
  const q = query.trim();
  if (!q) return [];
  const out: SearchResult[] = [];
  for (const d of corpus) {
    const ranges = findRanges(d.title, q);
    if (ranges.length === 0) continue;
    const earliest = Math.min(...ranges.map((r) => r[0]));
    // 越多命中 / 越靠前得分越高；常数 10 让命中数主导。
    const score =
      ranges.length * 10 + (1 - earliest / Math.max(d.title.length, 1));
    out.push(toSearchResult(d, score, ranges));
  }
  return out.sort((a, b) => b.score - a.score);
}

// ---------------- vector mode ----------------

function normalize(s: string): string {
  return s.toLowerCase().replace(/\s+/g, " ").trim();
}

function bigrams(s: string): string[] {
  const norm = normalize(s);
  if (norm.length === 0) return [];
  if (norm.length === 1) return [norm];
  const out: string[] = [];
  for (let i = 0; i < norm.length - 1; i++) {
    out.push(norm.slice(i, i + 2));
  }
  return out;
}

function termFreq(tokens: string[]): Map<string, number> {
  const tf = new Map<string, number>();
  for (const t of tokens) tf.set(t, (tf.get(t) ?? 0) + 1);
  return tf;
}

function buildIdf(docs: readonly { titleTokens: string[] }[]): Map<string, number> {
  const n = docs.length || 1;
  const df = new Map<string, number>();
  for (const d of docs) {
    const seen = new Set(d.titleTokens);
    for (const t of seen) df.set(t, (df.get(t) ?? 0) + 1);
  }
  const idf = new Map<string, number>();
  for (const [t, c] of df) idf.set(t, Math.log(1 + n / c));
  return idf;
}

function tfidfVec(tokens: string[], idf: Map<string, number>): Map<string, number> {
  const tf = termFreq(tokens);
  const total = tokens.length || 1;
  const v = new Map<string, number>();
  for (const [t, c] of tf) {
    const w = (c / total) * (idf.get(t) ?? 0);
    if (w > 0) v.set(t, w);
  }
  return v;
}

function cosine(a: Map<string, number>, b: Map<string, number>): number {
  const [small, large] = a.size <= b.size ? [a, b] : [b, a];
  let dot = 0;
  for (const [k, va] of small) {
    const vb = large.get(k);
    if (vb !== undefined) dot += va * vb;
  }
  if (dot === 0) return 0;
  let na = 0;
  let nb = 0;
  for (const v of a.values()) na += v * v;
  for (const v of b.values()) nb += v * v;
  if (na === 0 || nb === 0) return 0;
  return dot / (Math.sqrt(na) * Math.sqrt(nb));
}

function searchVector(query: string, corpus: readonly Doc[]): SearchResult[] {
  const q = query.trim();
  if (!q || corpus.length === 0) return [];
  const indexedDocs = corpus.map((doc) => ({
    ...doc,
    titleTokens: bigrams(doc.title),
  }));
  const idf = buildIdf(indexedDocs);
  const qVec = tfidfVec(bigrams(q), idf);
  if (qVec.size === 0) return [];
  const out: SearchResult[] = [];
  for (const d of indexedDocs) {
    const dVec = tfidfVec(d.titleTokens, idf);
    const score = cosine(qVec, dVec);
    if (score <= 0) continue;
    out.push(toSearchResult(d, score, []));
  }
  return out.sort((a, b) => b.score - a.score);
}

function buildIndexedCorpus(docs: Doc[], key: string): IndexedCorpus {
  const prepared = docs.map((doc) => ({
    ...doc,
    titleTokens: bigrams(doc.title),
    vector: new Map<string, number>(),
  }));
  const idf = buildIdf(prepared);
  for (const doc of prepared) {
    doc.vector = tfidfVec(doc.titleTokens, idf);
  }
  return {
    key,
    docs: prepared,
    idf,
  };
}

function getIndexedSidebarCorpus(): IndexedCorpus | null {
  const key = buildSidebarCorpusKey();
  if (!key) {
    indexedSidebarCorpus = null;
    return null;
  }
  if (indexedSidebarCorpus?.key === key) return indexedSidebarCorpus;
  indexedSidebarCorpus = buildIndexedCorpus(buildSidebarCorpus(), key);
  return indexedSidebarCorpus;
}

function searchVectorIndexed(query: string, corpus: IndexedCorpus): SearchResult[] {
  const q = query.trim();
  if (!q || corpus.docs.length === 0) return [];
  const qVec = tfidfVec(bigrams(q), corpus.idf);
  if (qVec.size === 0) return [];
  const out: SearchResult[] = [];
  for (const d of corpus.docs) {
    const score = cosine(qVec, d.vector);
    if (score <= 0) continue;
    out.push(toSearchResult(d, score, []));
  }
  return out.sort((a, b) => b.score - a.score);
}

// ---------------- merge ----------------

/** 合并两路结果：text 分数除以最高 text 分归一到 [0,1]；权重 0.6/0.4 略偏向文本。 */
function searchMerged(query: string, textHits: SearchResult[], vectorHits: SearchResult[]): SearchResult[] {
  const q = query.trim();
  if (!q) return [];
  const TEXT_WEIGHT = 0.6;
  const VECTOR_WEIGHT = 0.4;

  const textMax = textHits.length ? textHits[0].score : 0;
  const norm = (s: number) => (textMax > 0 ? s / textMax : 0);

  const merged = new Map<string, SearchResult>();

  for (const r of textHits) {
    merged.set(r.route, {
      ...r,
      score: norm(r.score) * TEXT_WEIGHT,
    });
  }

  for (const r of vectorHits) {
    const prev = merged.get(r.route);
    if (prev) {
      merged.set(r.route, {
        ...prev,
        score: prev.score + r.score * VECTOR_WEIGHT,
      });
    } else {
      merged.set(r.route, {
        ...r,
        score: r.score * VECTOR_WEIGHT,
      });
    }
  }

  return Array.from(merged.values()).sort((a, b) => b.score - a.score);
}

// ---------------- public ----------------

export function searchSessions(query: string): SearchResult[] {
  const sidebarCorpus = getIndexedSidebarCorpus();
  if (sidebarCorpus) {
    return searchMerged(
      query,
      searchText(query, sidebarCorpus.docs),
      searchVectorIndexed(query, sidebarCorpus),
    );
  }
  const storeCorpus = buildStoreCorpus();
  return searchMerged(
    query,
    searchText(query, storeCorpus),
    searchVector(query, storeCorpus),
  );
}

export async function ensureSessionSearchCorpusLoaded(force = false): Promise<void> {
  await ensureSidebarConversationsLoaded(force);
}

File: apps/desktop/src/application/session_search.rs
//! Session title search shared by desktop consumers.
//!
//! Substring hits rank first, followed by character
//! bigram TF-IDF + cosine similarity for near matches. Corpus is derived from
//! Product Core task facts — never from UI caches.

use std::collections::{BTreeMap, HashMap, HashSet};
use std::hash::{Hash, Hasher};
use std::sync::{Arc, Mutex};

use lilia_contracts::{ProductTask, Project, ProjectId, TaskId};
use serde::{Deserialize, Serialize};

use crate::application::{DesktopApplication, DesktopApplicationError, ProjectQuery, TaskQuery};

const MAX_SESSION_SEARCH_RESULTS: usize = 100;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum DesktopSessionSearchKind {
    ProjectTask,
    Orphan,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DesktopSessionSearchResult {
    pub kind: DesktopSessionSearchKind,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub project_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub project_name: Option<String>,
    pub task_id: TaskId,
    pub title: String,
    pub route: String,
    pub score: f64,
    pub highlights: Vec<(usize, usize)>,
}

#[derive(Clone)]
struct SessionDoc {
    kind: DesktopSessionSearchKind,
    project_id: Option<String>,
    project_name: Option<String>,
    task_id: TaskId,
    title: String,
    route: String,
    title_tokens: Vec<String>,
    vector: HashMap<String, f64>,
}

/// 语料的 bigram 与 IDF 只依赖任务事实。指纹相同就复用，避免每次搜索都重建。
/// 用指纹而不是事件失效：漏一次订阅会让搜索结果长期陈旧，指纹不会。
struct SessionSearchCorpus {
    fingerprint: u64,
    docs: Vec<SessionDoc>,
    idf: HashMap<String, f64>,
}

fn corpus_fingerprint(projects: &[Project], tasks: &[ProductTask]) -> u64 {
    let mut hasher = std::collections::hash_map::DefaultHasher::new();
    for project in projects {
        project.id.as_str().hash(&mut hasher);
        project.name.hash(&mut hasher);
    }
    for task in tasks {
        task.id.as_str().hash(&mut hasher);
        task.title.hash(&mut hasher);
        task.project_id
            .as_ref()
            .map(ProjectId::as_str)
            .hash(&mut hasher);
    }
    hasher.finish()
}

#[derive(Clone)]
pub struct SessionSearchService {
    projects: lilia_feature_task::ProjectTaskService,
    corpus: Arc<Mutex<Option<Arc<SessionSearchCorpus>>>>,
}
impl SessionSearchService {
    pub(crate) fn new(projects: lilia_feature_task::ProjectTaskService) -> Self {
        Self {
            projects,
            corpus: Arc::new(Mutex::new(None)),
        }
    }

    pub fn search_sessions(
        &self,
        query: &str,
        limit: usize,
    ) -> Result<Vec<DesktopSessionSearchResult>, DesktopApplicationError> {
        let query = query.trim();
        if query.is_empty() {
            return Ok(Vec::new());
        }
        let corpus = self.session_search_corpus()?;
        let mut merged = BTreeMap::new();
        for result in search_text(query, &corpus.docs) {
            merged.insert(result.route.clone(), result);
        }
        for result in search_vector(query, &corpus.docs, &corpus.idf) {
            merged
                .entry(result.route.clone())
                .and_modify(|existing| {
                    if result.score > existing.score {
                        existing.score = result.score;
                    }
                    if existing.highlights.is_empty() && !result.highlights.is_empty() {
                        existing.highlights = result.highlights.clone();
                    }
                })
                .or_insert(result);
        }
        let mut results = merged.into_values().collect::<Vec<_>>();
        results.sort_by(|left, right| {
            right
                .score
                .partial_cmp(&left.score)
                .unwrap_or(std::cmp::Ordering::Equal)
                .then_with(|| left.title.cmp(&right.title))
                .then_with(|| left.task_id.as_str().cmp(right.task_id.as_str()))
        });
        results.truncate(limit.clamp(1, MAX_SESSION_SEARCH_RESULTS));
        Ok(results)
    }

    fn session_search_corpus(&self) -> Result<Arc<SessionSearchCorpus>, DesktopApplicationError> {
        let projects = self.projects.query_projects(ProjectQuery::default())?;
        let tasks = self.projects.query_tasks(TaskQuery::default())?;
        let fingerprint = corpus_fingerprint(&projects, &tasks);
        if let Some(cached) = self
            .corpus
            .lock()
            .unwrap_or_else(|error| error.into_inner())
            .as_ref()
            .filter(|cached| cached.fingerprint == fingerprint)
        {
            return Ok(Arc::clone(cached));
        }
        let project_names = projects
            .into_iter()
            .map(|project| (project.id, project.name))
            .collect::<BTreeMap<_, _>>();
        let mut docs = Vec::new();
        for task in tasks {
            let project_name = task
                .project_id
                .as_ref()
                .and_then(|project_id| project_names.get(project_id))
                .cloned();
            let project_id = task.project_id.as_ref().map(|id| id.as_str().to_owned());
            let route = match &project_id {
                Some(project_id) => format!("/projects/{project_id}/tasks/{}", task.id.as_str()),
                None => format!("/chats/{}", task.id.as_str()),
            };
            let title_tokens = bigrams(&task.title);
            docs.push(SessionDoc {
                kind: if project_id.is_some() {
                    DesktopSessionSearchKind::ProjectTask
                } else {
                    DesktopSessionSearchKind::Orphan
                },
                project_id,
                project_name,
                task_id: task.id,
                title: task.title,
                route,
                title_tokens,
                vector: HashMap::new(),
            });
        }
        let idf = build_idf(&docs);
        for doc in &mut docs {
            doc.vector = tfidf_vec(&doc.title_tokens, &idf);
        }
        let corpus = Arc::new(SessionSearchCorpus {
            fingerprint,
            docs,
            idf,
        });
        *self
            .corpus
            .lock()
            .unwrap_or_else(|error| error.into_inner()) = Some(Arc::clone(&corpus));
        Ok(corpus)
    }
}

pub enum SessionSearchServiceKey {}
impl lilia_kernel::ServiceKey for SessionSearchServiceKey {
    type Value = SessionSearchService;
    const NAME: &'static str = "lilia.session.search";
}
pub struct SessionSearchFeature {
    service: SessionSearchService,
}
impl SessionSearchFeature {
    pub fn new(service: SessionSearchService) -> Self {
        Self { service }
    }
}
impl lilia_kernel::Feature for SessionSearchFeature {
    fn id(&self) -> lilia_kernel::FeatureId {
        lilia_kernel::FeatureId::new("lilia.feature.session-search").expect("valid feature id")
    }
    fn provides(&self) -> Vec<lilia_kernel::ServiceRef> {
        vec![lilia_kernel::ServiceRef::of::<SessionSearchServiceKey>()]
    }
    fn mount(
        &self,
        cx: &mut lilia_kernel::FeatureContext<'_>,
    ) -> Result<(), lilia_kernel::KernelError> {
        cx.provide::<SessionSearchServiceKey>(self.service.clone())
    }
}
impl DesktopApplication {
    pub fn session_search_service(&self) -> SessionSearchService {
        self.inner.session_search.clone()
    }
    pub fn search_sessions(
        &self,
        query: &str,
        limit: usize,
    ) -> Result<Vec<DesktopSessionSearchResult>, DesktopApplicationError> {
        self.inner.session_search.search_sessions(query, limit)
    }
}

fn search_text(query: &str, docs: &[SessionDoc]) -> Vec<DesktopSessionSearchResult> {
    let mut out = Vec::new();
    for doc in docs {
        let ranges = find_ranges(&doc.title, query);
        if ranges.is_empty() {
            continue;
        }
        let earliest = ranges.iter().map(|range| range.0).min().unwrap_or(0);
        let score =
            ranges.len() as f64 * 10.0 + (1.0 - earliest as f64 / doc.title.len().max(1) as f64);
        out.push(to_result(doc, score, ranges));
    }
    out
}

fn search_vector(
    query: &str,
    docs: &[SessionDoc],
    idf: &HashMap<String, f64>,
) -> Vec<DesktopSessionSearchResult> {
    let query_vec = tfidf_vec(&bigrams(query), idf);
    if query_vec.is_empty() {
        return Vec::new();
    }
    let mut out = Vec::new();
    for doc in docs {
        let score = cosine(&query_vec, &doc.vector);
        if score <= 0.05 {
            continue;
        }
        out.push(to_result(doc, score, Vec::new()));
    }
    out
}

fn to_result(
    doc: &SessionDoc,
    score: f64,
    highlights: Vec<(usize, usize)>,
) -> DesktopSessionSearchResult {
    DesktopSessionSearchResult {
        kind: doc.kind,
        project_id: doc.project_id.clone(),
        project_name: doc.project_name.clone(),
        task_id: doc.task_id.clone(),
        title: doc.title.clone(),
        route: doc.route.clone(),
        score,
        highlights,
    }
}

fn find_ranges(title: &str, query: &str) -> Vec<(usize, usize)> {
    let title_lower = title.to_lowercase();
    let query_lower = query.to_lowercase().trim().to_owned();
    if query_lower.is_empty() {
        return Vec::new();
    }
    let mut ranges = Vec::new();
    let mut idx = 0;
    while let Some(found) = title_lower[idx..].find(&query_lower) {
        let start = idx + found;
        let end = start + query_lower.len();
        ranges.push((start, end));
        idx = end;
    }
    if !ranges.is_empty() {
        return ranges;
    }
    for token in query_lower
        .split_whitespace()
        .filter(|token| !token.is_empty())
    {
        let mut i = 0;
        while let Some(found) = title_lower[i..].find(token) {
            let start = i + found;
            let end = start + token.len();
            ranges.push((start, end));
            i = end;
        }
    }
    ranges
}

fn normalize(value: &str) -> String {
    value
        .to_lowercase()
        .split_whitespace()
        .collect::<Vec<_>>()
        .join(" ")
}

fn bigrams(value: &str) -> Vec<String> {
    let norm = normalize(value);
    if norm.is_empty() {
        return Vec::new();
    }
    let chars = norm.chars().collect::<Vec<_>>();
    if chars.len() == 1 {
        return vec![norm];
    }
    chars
        .windows(2)
        .map(|window| window.iter().collect::<String>())
        .collect()
}

fn build_idf(docs: &[SessionDoc]) -> HashMap<String, f64> {
    let n = docs.len().max(1) as f64;
    let mut df = HashMap::new();
    for doc in docs {
        let seen = doc.title_tokens.iter().cloned().collect::<HashSet<_>>();
        for token in seen {
            *df.entry(token).or_insert(0_u64) += 1;
        }
    }
    df.into_iter()
        .map(|(token, count)| (token, (1.0 + n / count as f64).ln()))
        .collect()
}

fn tfidf_vec(tokens: &[String], idf: &HashMap<String, f64>) -> HashMap<String, f64> {
    let mut tf = HashMap::new();
    for token in tokens {
        *tf.entry(token.clone()).or_insert(0_u64) += 1;
    }
    let total = tokens.len().max(1) as f64;
    tf.into_iter()
        .filter_map(|(token, count)| {
            let weight = (count as f64 / total) * idf.get(&token).copied().unwrap_or(0.0);
            (weight > 0.0).then_some((token, weight))
        })
        .collect()
}

fn cosine(left: &HashMap<String, f64>, right: &HashMap<String, f64>) -> f64 {
    if left.is_empty() || right.is_empty() {
        return 0.0;
    }
    let mut dot = 0.0;
    for (token, left_weight) in left {
        if let Some(right_weight) = right.get(token) {
            dot += left_weight * right_weight;
        }
    }
    let left_norm = left.values().map(|value| value * value).sum::<f64>().sqrt();
    let right_norm = right
        .values()
        .map(|value| value * value)
        .sum::<f64>()
        .sqrt();
    if left_norm == 0.0 || right_norm == 0.0 {
        0.0
    } else {
        dot / (left_norm * right_norm)
    }
}

#[cfg(test)]
mod tests {
    use std::sync::Arc;

    use tempfile::TempDir;
    use uuid::Uuid;

    use super::*;
    use crate::application::{
        DesktopApplicationConfig, DesktopHost, DesktopHostAction, DesktopHostContext,
        DesktopHostError, DesktopHostResult, DesktopProjectCreate, DesktopTaskCreate,
    };

    #[derive(Default)]
    struct TestHost;

    impl DesktopHost for TestHost {
        fn execute(
            &self,
            _context: &DesktopHostContext,
            _action: DesktopHostAction,
        ) -> Result<DesktopHostResult, DesktopHostError> {
            Ok(DesktopHostResult::Completed)
        }
    }

    fn temp_app() -> (TempDir, DesktopApplication) {
        let dir = TempDir::new().unwrap();
        let config =
            DesktopApplicationConfig::new(dir.path(), format!("search-{}", Uuid::new_v4()))
                .unwrap();
        let app = DesktopApplication::bootstrap(config, Arc::new(TestHost)).unwrap();
        (dir, app)
    }

    #[test]
    fn substring_hit_ranks_above_unrelated_titles() {
        let (_dir, app) = temp_app();
        let project = app
            .create_project(DesktopProjectCreate::new("Alpha"))
            .unwrap();
        let matching = app
            .create_task(DesktopTaskCreate::new(
                Some(project.id.clone()),
                "Fix login timeout",
            ))
            .unwrap();
        app.create_task(DesktopTaskCreate::new(
            Some(project.id.clone()),
            "Unrelated gardening notes",
        ))
        .unwrap();

        let results = app.search_sessions("login", 10).unwrap();
        assert!(!results.is_empty());
        assert_eq!(results[0].task_id, matching.id);
        assert!(!results[0].highlights.is_empty());
        assert!(results.iter().all(|result| result.title.contains("login")
            || result.score > 0.0 && result.task_id == matching.id));
    }

    #[test]
    fn direct_search_refreshes_renamed_and_archived_facts_without_a_subscription() {
        let (_dir, app) = temp_app();
        let project = app
            .create_project(DesktopProjectCreate::new("Original"))
            .unwrap();
        let task = app
            .create_task(DesktopTaskCreate::new(
                Some(project.id.clone()),
                "unique needle",
            ))
            .unwrap();
        let service = app.session_search_service();
        assert_eq!(
            service.search_sessions("needle", 10).unwrap()[0]
                .project_name
                .as_deref(),
            Some("Original")
        );
        app.inner
            .project_tasks
            .update_project(
                &project.id,
                crate::application::DesktopProjectPatch {
                    name: Some("Renamed".into()),
                    ..Default::default()
                },
            )
            .unwrap();
        app.inner
            .project_tasks
            .update_task(
                &task.id,
                crate::application::DesktopTaskPatch {
                    title: Some("unique changed".into()),
                    ..Default::default()
                },
            )
            .unwrap();
        let result = service.search_sessions("changed", 10).unwrap();
        assert_eq!(result[0].task_id, task.id);
        assert_eq!(result[0].title, "unique changed");
        assert_eq!(result[0].project_name.as_deref(), Some("Renamed"));
        app.inner
            .project_tasks
            .set_task_archived(&task.id, true)
            .unwrap();
        assert!(service.search_sessions("changed", 10).unwrap().is_empty());
    }

    #[test]
    fn empty_query_returns_no_results() {
        let (_dir, app) = temp_app();
        assert!(app.search_sessions("   ", 10).unwrap().is_empty());
    }
}
